### PR TITLE
Receipts optimization

### PR DESCRIPTION
**Pull request by sasurobert**
Added new field to block header to keep the receipts only for in a single hash.

### CODE REVIEW

File: process/block/baseProcess.go [hunk 1]
```diff
@@ -1131,3 +1131,26 @@ func (bp *baseProcessor) commitAll() error {
...
+			continue
+		}
+
+		if mb.Type != block.ReceiptBlock && mb.Type != block.SmartContractResultBlock {
+			i++
+			continue
+		}
+
+		body[i] = body[len(body)-1]
+		body = body[:len(body)-1]
```
Comment (iulianpascalau) [inline process/block/baseProcess.go]: Seems legit now after the talk. :+1:

File: process/block/metablock_test.go [hunk 2]
```diff
@@ -839,7 +839,7 @@ func TestMetaProcessor_ApplyBodyToHeaderShouldWork(t *testing.T) {
 	mp, _ := blproc.NewMetaProcessor(arguments)
 
 	hdr := &block.MetaBlock{}
-	err := mp.ApplyBodyToHeader(hdr, nil)
+	_, err := mp.ApplyBodyToHeader(hdr, nil)
```
Comment (iulianpascalau) [inline process/block/metablock_test.go]: Do not forget to remove comments on lines 643 and 656 in transactions.go (//what is this?) :D

File: process/block/postprocess/basePostProcess.go [hunk 3]
```diff
@@ -130,3 +131,10 @@ func (bpp *basePostProcessor) verifyMiniBlock(createMBs map[uint32]*block.MiniBl
 
 	return nil
 }
+
+func (bpp *basePostProcessor) GetCreatedInShardMiniBlock() *block.MiniBlock {
```
Comment (iulianpascalau) [inline process/block/postprocess/basePostProcess.go]: Please correct the log on line 57 UnsignedTransactionUnit.Put -> something else involving bpp.storageType

File: process/block/shardblock.go [hunk 4]
```diff
@@ -1637,18 +1637,26 @@ func (sp *shardProcessor) ApplyBodyToHeader(hdr data.HeaderHandler, bodyHandler
 		go sp.checkAndRequestIfMetaHeadersMissing(hdr.GetRound())
 	}()
 
-	if bodyHandler == nil || bodyHandler.IsInterfaceNil() {
-		return nil
+	if check.IfNil(bodyHandler) {
+		return bodyHandler, nil
```
Comment (iulianpascalau) [inline process/block/shardblock.go]: we are returning nil, nil here. Why not returning an error?

File: process/coordinator/process.go [hunk 5]
```diff
@@ -740,12 +750,42 @@ func (tc *transactionCoordinator) VerifyCreatedBlockTransactions(body block.Body
...
+		currHash, err := core.CalculateHash(tc.marshalizer, tc.hasher, mb)
+		if err != nil {
+			return nil, err
+		}
+
+		allReceiptsHashes = append(allReceiptsHashes, currHash...)
 	}
 
-	return interimProc.VerifyInterMiniBlocks(body)
+	finalReceiptHash, err := core.CalculateHash(tc.marshalizer, tc.hasher, allReceiptsHashes)
```
Comment (iulianpascalau) [inline process/coordinator/process.go]: eth uses a merkle trie for this. :)

File: integrationTests/multiShard/block/executingMiniblocks_test.go [hunk 6]
```diff
@@ -280,8 +281,29 @@ func TestSimpleTransactionsWithMoreValueThanBalanceYieldReceiptsInMultiShardedEn
...
+			if mb.Type == block.InvalidBlock {
+				numInvalid++
+			}
+		}
+		assert.Equal(t, 1, numInvalid)
+	}
+
 	time.Sleep(time.Second)
-	nrRoundsToTest := 10
+	nrRoundsToTest := 6
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/multiShard/block/executingMiniblocks_test.go]: done.

File: process/block/metablock_test.go (hunk 2 above)
Comment (sasurobert) [reply process/block/metablock_test.go]: done.

File: process/block/postprocess/basePostProcess.go (hunk 3 above)
Comment (sasurobert) [reply process/block/postprocess/basePostProcess.go]: done.

File: process/block/shardblock.go (hunk 4 above)
Comment (sasurobert) [reply process/block/shardblock.go]: done.

**Pull request merged**